lib/cc/analyzer/include_paths_builder.rb

```diff
@@ -0,0 +1,140 @@
+module CC
+  module Analyzer
+    class IncludePathsBuilder
+      def initialize(cc_excludes)
+        @cc_excludes = cc_excludes
+      end
+
+      def build
+        root = Directory.new('.', included_files)
+        root.included_paths
+      end
+
+      protected
+
+      def ignored_files
+        Tempfile.open(".cc_gitignore") do |tmp|
+          tmp.write(File.read(".gitignore")) if File.exist?(".gitignore")
+          tmp << @cc_excludes.join("\n")
+          tmp.close
+          tracked_and_ignored = `git ls-files -zi -X #{tmp.path}`.split("\0")
+          untracked_and_ignored = `git ls-files -zio -X #{tmp.path}`.split("\0")
+          tracked_and_ignored + untracked_and_ignored
+        end
+      end
+
+      def included_files
+        includable_files - ignored_files
+      end
+
+      def includable_files
+        tracked_in_git = `git ls-files -z`.split("\0")
+        untracked_in_git = `git ls-files -zo`.split("\0")
+        tracked_in_git + untracked_in_git
+      end
+
+      class Directory
+        def initialize(path, included_files)
+          @path = path
+          @included_files = ensure_hashified(included_files)
+        end
+
+        def all_included?
+          readable? && files_all_included? && subdirectories_all_included?
+        end
+
+        def included_paths
+          if all_included?
+            [@path + "/"]
+          elsif readable?
+            result = []
+            result = result.concat(included_file_entries)
+            result = result.concat(included_subdirectory_results)
+            result
+          else
+            []
+          end
+        end
+
+        protected
+
+        def ensure_hashified(obj)
+          if obj.is_a?(Array)
+            result = {}
+            obj.each do |included|
+              result[included] = true
+            end
+            result
+          else
+            obj
+          end
+        end
+
+        def files_all_included?
+          file_entries.all? { |e| @included_files[e] }
+        end
+
+        def file_entries
+          unless @_file_entries
+            @_file_entries = relevant_full_entries.reject do |e|
+              File.directory?(e)
+            end
+          end
+          @_file_entries
+        end
+
+        def included_file_entries
+          file_entries.select { |file_entry| @included_files[file_entry] }
+        end
+
+        def included_subdirectory_results
+          result = []
+          subdirectories.each do |subdirectory|
+            result = result.concat(subdirectory.included_paths)
+          end
+          result
+        end
+
+        def readable?
```

Comment:
Originally the requirements of this solution were such that, if there were files/directories that were not readable and weren't intended to be excluded, those were still supposed to bubble up as permissions errors.

It was only when the engines encountered unreadable files *that the user had intended to exclude* that is "the permissions  bug" we're trying to fix.

It might be a fine 95% solution, or the best we can do given all the subtle hurdles we've encountered, but if the requirement were boiled down to: "just skip unreadable stuff", there may have been something simpler that didn't involve all the git tom-foolery:

- Recursively expand the source dir
- Rescue permissions errors and ignore
- Filter excluded
- Present the result as `include_paths`

Right?

Also, we have to keep in mind that the CLI (this code) runs as root (technically whatever user is UID:0), while the engines run as an underprivileged user (app:9000), so anything you do permissions-related in this code isn't guaranteed to translate to what the engine is ultimately able to access.